packages/e2e-test-utils-playwright/src/page-utils/press-keys.ts

```diff
@@ -47,64 +47,157 @@ export function setClipboardData(
+	// This is needed for the `paste` event to be fired in case of a real key press.
+	await this.page.evaluate( ( data ) => {
+		const activeElement =
+			document.activeElement instanceof HTMLIFrameElement
+				? document.activeElement.contentDocument!.activeElement
+				: document.activeElement;
+		activeElement?.addEventListener(
+			'copy',
+			( event ) => {
+				event.preventDefault();
+				event.stopImmediatePropagation();
+				Object.entries( data ).forEach( ( [ type, text ] ) => {
+					if ( text ) {
+						( event as ClipboardEvent ).clipboardData?.setData(
+							type,
+							text
+						);
+					}
+				} );
+			},
+			{ once: true, capture: true }
+		);
+	}, clipboardDataHolder );
+
+	await this.page.keyboard.press(
+		isAppleOS() ? 'Meta+KeyC' : 'Control+KeyC'
+	);
 }
 
 async function emulateClipboard( page: Page, type: 'copy' | 'cut' | 'paste' ) {
-	clipboardDataHolder = await page.evaluate(
+	const promiseHandle = await page.evaluateHandle(
 		( [ _type, _clipboardData ] ) => {
-			const canvasDoc =
-				// @ts-ignore
-				document.activeElement?.contentDocument ?? document;
-			const clipboardDataTransfer = new DataTransfer();
-
-			if ( _type === 'paste' ) {
-				clipboardDataTransfer.setData(
-					'text/plain',
-					_clipboardData[ 'text/plain' ]
-				);
-				clipboardDataTransfer.setData(
-					'text/html',
-					_clipboardData[ 'text/html' ]
-				);
-				clipboardDataTransfer.setData(
-					'rich-text',
-					_clipboardData[ 'rich-text' ]
-				);
-			} else {
-				const selection = canvasDoc.defaultView.getSelection()!;
-				const plainText = selection.toString();
-				let html = plainText;
-				if ( selection.rangeCount ) {
-					const range = selection.getRangeAt( 0 );
-					const fragment = range.cloneContents();
-					html = Array.from( fragment.childNodes )
-						.map(
-							( node ) =>
-								( node as Element ).outerHTML ??
-								( node as Element ).nodeValue
-						)
-						.join( '' );
-				}
-				clipboardDataTransfer.setData( 'text/plain', plainText );
-				clipboardDataTransfer.setData( 'text/html', html );
-			}
-
-			canvasDoc.activeElement?.dispatchEvent(
-				new ClipboardEvent( _type, {
-					bubbles: true,
-					cancelable: true,
-					clipboardData: clipboardDataTransfer,
-				} )
-			);
+			const activeElement =
+				document.activeElement instanceof HTMLIFrameElement
+					? document.activeElement.contentDocument!.activeElement
+					: document.activeElement;
 
+			// Return an object with the promise handle to bypass the auto-resolving
+			// feature of `evaluateHandle()`.
 			return {
-				'text/plain': clipboardDataTransfer.getData( 'text/plain' ),
-				'text/html': clipboardDataTransfer.getData( 'text/html' ),
-				'rich-text': clipboardDataTransfer.getData( 'rich-text' ),
+				promise: new Promise< false | typeof clipboardDataHolder >(
+					( resolve ) => {
+						const timeout = setTimeout( () => {
+							resolve( false );
+						}, 50 );
```

Comment:
No, this is here in case the event never fires. This could happen for `paste` events when the element doesn't attach an event listener and it needs to fallback to native "real" event. We need a flag here that basically means `hasEventHandled`, and it that's `false` we'll fire a real key press.

I know this is all very confusing and I've been thinking about better ways to document this 😅. Maybe I should make a comparison table or something like that...

Reply:
Any additional documentation would be helpful here.